src/contract/event.rs

```diff
@@ -516,12 +544,103 @@ impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
+    /// single query.
+    ///
+    /// # Panics
+    ///
+    /// If the block range is invalid for querying past events:
+    /// - the from block is "latest" or "pending"
+    /// - the to block is "earliest"
+    pub async fn query_past_events_paginated(self) -> Result<Vec<Event<E>>, ExecutionError> {
+        let mut start_block = match self.from_block {
+            None | Some(BlockNumber::Earliest) => 0,
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                panic!("invalid from block value 'latest' or 'pending'")
+            }
+        };
+        if let Some(deployment_tx) = self.deployment_transaction {
+            start_block = cmp::max(
+                start_block,
+                block_number_from_transaction_hash(self.web3.clone(), deployment_tx).await?,
+            );
+        }
+
+        let end_block = match self.to_block {
+            None | Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                self.web3.eth().block_number().compat().await?.as_u64()
+            }
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Earliest) => panic!("invalid to block value 'earliest'"),
+        };
+        let page_size = self.block_page_size.unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+        let (web3, filter) = self.prepare();
+
+        let mut current_block = start_block;
+        let mut events = Vec::new();
+        let mut append_events = |logs: Vec<Log>| -> Result<(), ExecutionError> {
+            events.reserve(logs.len());
+            for log in logs {
+                let event = Event::from_log(log, E::parse_log)?;
+                events.push(event);
+            }
+            Ok(())
+        };
+
+        // NOTE: Query only until the last page, since it is handled a little
+        //   differently to deal with "latest" and "pending" to blocks.
+        while current_block + page_size <= end_block {
+            let filter = filter
+                .clone()
+                .from_block(current_block.into())
+                .to_block((current_block + page_size - 1).into())
+                .build();
+            let event_page = web3.eth().logs(filter).compat().await?;
+            append_events(event_page)?;
+
+            current_block += page_size;
+        }
+
+        // NOTE: In case this was called with a from block of "latest" or
+        //   "pending" we want to make sure that the last call includes blocks
+        //   that have been added since the start of the call. To do this, we
+        //   just omit the `to_block` from the `filter` since it was already
+        //   specified in the `prepare` method. In the case a block number was
+        //   specified, this call is still correct as the filter's `to_block`
+        //   will be set to that block number.
+        if start_block > end_block {
```

Comment:
I think I understand the comment but why are we doing this comparison? Shouldn't we check whether to block was Latest or Pending and ignore start block?

Reply:
You're right, this comparison is a little confusing, I had to explain it when were pair programming as well as it wan't obvious. I can move it.

Essentially the for loop queries all but the last page, the last page is queried with the `to_block` that was specified by the builder.

The comparison is just in case the start block was after the end block in which case we should skip everything but I think the comparison should be moved further up.

Reply:
No, because the filter has the `from_block` and `to_block` set from the builder. I think this needs to be a bit more explicit.

Reply:
Shouldn't it be ` if start_block < end_block` instead of `>` then? We skip everything if the range between the blocks empty.

Reply:
I updated this and the comment, @fleupold and @e00E can you guys let me know if it makes more sense now?

Reply:
Makes sense to me.